### PR TITLE
feat: Add `Tree` components

### PR DESCRIPTION
**Pull request by natanelia**
Add `Tree` components according to WAI-ARIA TreeView declared properties specification.

Link:
https://www.w3.org/TR/wai-aria-practices-1.2/examples/treeview/treeview-1/treeview-1b.html

- API Design: https://gist.github.com/diegohaz/6bfdc758cf84dcff32f26b91e5c2f933
- Status: OK to Release
- Todos:
   - [x] Make it works
   - [x] Make it WAI-ARIA compliant
      - [x] Screen readers
      - [x] Keyboard navigation
   - [x] Create all tests
   - [x] Style example properly
   - [ ] Performance optimisation

### CODE REVIEW

**Comment by diegohaz** [comment]
Hey @natanelia! Thank you so much for working on this.

I still haven't looked deep into the code, but testing the component, I noticed some bugs or missing features:

1. Pressing the arrow right on an expanded item should focus on the first item in the nested group, but nothing happens.
2. When I press the arrow right on an expanded item, as stated above, nothing happens, but then I have to press the arrow left a few times to collapse the group again.
3. Pressing the arrow left on a nested item should focus on the parent item, but it doesn't work.

Demonstrating the bugs or missing features on the tree component

Regarding the API, I love that you could implement that in the gist. Do you think that API makes sense?

I just noticed that `expandedIds` refer to the label IDs, not the item ids. From an API perspective, I think it would make more sense to use only the IDs passed to `TreeItem`.

Finally, I'd like to support a flat API at some point:

```jsx
<Tree state={tree}>
  <TreeItem id="item-1">Item 1</TreeItem>
  <TreeItem groupId="item-1">Item 1.1</TreeItem>
  <TreeItem groupId="item-1">Item 1.2</TreeItem>
  <TreeItem groupId="item-1">Item 1.3</TreeItem>
  <TreeItem id="item-2">Item 2</TreeItem>
  <TreeItem groupId="item-2">Item 2.1</TreeItem>
  <TreeItem groupId="item-2" id="item-2-2">Item 2.2</TreeItem>
  <TreeItem groupId="item-2-2">Item 2.2.1</TreeItem>
  <TreeItem groupId="item-2-2">Item 2.2.2</TreeItem>
  <TreeItem groupId="item-2-2">Item 2.2.3</TreeItem>
  <TreeItem groupId="item-2">Item 2.3</TreeItem>
</Tree>
```

The main reason is that it's much easier to implement virtualized lists with a flat tree like that. That's how VSCode renders its file tree, for example:

Screenshot of VSCode with the elements inspector open focusing on a tree item element

**Comment by natanelia** [comment]
Hi @diegohaz thanks for the feedbacks!

> I still haven't looked deep into the code, but testing the component, I noticed some bugs or missing features

Thank you for looking into it. I'll fix the bugs soon with the tests.

---
> Regarding the API, I love that you could implement that in the gist. Do you think that API makes sense?

The API does reflect this [example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/treeview/treeview-1/treeview-1b.html) very well.

During development, I do encounter this problem, if we implement click and keyboard handlers to the TreeItem instead of the TreeGroupLabel, developer has to implement proper CSS or this will happen
    Screenshot 2022-02-18 at 5 14 46 PM

---
> Finally, I'd like to support a flat API at some point

I agree this will be a huge help for virtualising. For once, I thought the DOM structure is mandatory. There are these specifications that made me hesitate to make it flat-able.
- Each parent node contains or owns an element with role [group](https://www.w3.org/TR/wai-aria-1.2/#group).
- Each child node is contained in or owned by an element with role [group](https://www.w3.org/TR/wai-aria-1.2/#group) that is contained in or owned by the node that serves as the parent of that child.

The questions for virtualisation are:
- Is group actually not mandatory?
- If it is mandatory, what do you think is the best way to implement it? Do you think it is a good idea to nest invisible element inside TreeItem and use aria-owns to refer to each of its child?

**Comment by diegohaz** [comment]
I just tested the UI quickly, and it's working great. The issues I mentioned are fixed. Awesome work!

I'll take a better look into the code after wrapping up the work on Select. But after a glance, I wonder if we can simplify the state a little bit. Composite already uses Collection underneath. Do you think we can use the `tree.items` state instead of `tree.treeItems.items`?

Regarding the flat tree on SSR, the most appropriate solution I can think of is accepting a hardcoded prop (let's say `level`) on the TreeItem component. We do something similar on other components like Tab. The initially selected tab must be explicitly set on the tab state to work with SSR.

**Comment by natanelia** [comment]
> But after a glance, I wonder if we can simplify the state a little bit. Composite already uses Collection underneath. Do you think we can use the tree.items state instead of tree.treeItems.items?

We can if we override the disabled prop passed to the useCompositeItem to control focusability on non-expanded's child items. I will try this.

> Regarding the flat tree on SSR, the most appropriate solution I can think of is accepting a hardcoded prop (let's say level) on the TreeItem component. We do something similar on other components like Tab. The initially selected tab must be explicitly set on the tab state to work with SSR.

This is a good idea. However, we need a mechanism to control deeply nested tree items visibility. Otherwise, developers should define all of those items in "defaultExpandedIds" (which I don't think is a good DX).

For example we have this tree:
- A
  - A.1
     - A.1.1
        - A.1.1.1

And we have defaultExpandedIds only set on A.1.1.

In SSR, since A is not expanded, how do we make sure only A is rendered?

**Comment by diegohaz** [comment]
> We can if we override the disabled prop passed to the useCompositeItem to control focusability on non-expanded's child items. I will try this.

Can we set `shouldRegisterItem: false` when the items are collapsed?

> This is a good idea. However, we need a mechanism to control deeply nested tree items visibility. Otherwise, developers should define all of those items in "defaultExpandedIds" (which I don't think is a good DX).
> 
> For example we have this tree:
> 
> * A
>   
>   * A.1
>     
>     * A.1.1
>       
>       * A.1.1.1
> 
> And we have defaultExpandedIds only set on A.1.1.
> 
> In SSR, since A is not expanded, how do we make sure only A is rendered?

I don't know if it could be better, but that's actually a valid state. Visually speaking, `A.1.1` could be expanded while `A` is collapsed:

Expanding and collapsing a tree item with a nested expanded tree item

**Comment by natanelia** [comment]
I think there might be some implementation problem here. I'll try to answer based on current implementation, so we can see what can be better.

> Can we set shouldRegisterItem: false when the items are collapsed?

I don't think we can. Current useTreeItem depends on `childTreeItems` variable, which is used to determine whether the tree item has child and therefore can be expanded or not. Using shouldRegisterItem means those child will not be registered. So the parent won't be able to determine its expandability.

> I don't know if it could be better, but that's actually a valid state. Visually speaking, A.1.1 could be expanded while A is collapsed

Apologize, I updated my previous comment to remove the wrong idea.

Yes it is indeed a valid state. What I'm trying to say is this: To decide the visibility of A.1.1.1, it needs to know who are its ancestors (A, A.1, A.1.1) and their expanded state beforehand, during SSR. But the data is not available. Using state.items won't be enough, because the items are not registered during SSR. That is what TreeItemSyncContext is trying to solve.

If we are solving this using props, my assumption is that we are gonna write something like this (prop name "isParentExpandedAndVisible" is just an illustration):
```
const tree = useTreeState({ defaultExpandedIds: ['A.1.1'] });
...
<TreeItem id="A">A</TreeItem>
<TreeItem groupId="A" id="A.1" isParentExpandedAndVisible={false}>A.1</TreeItem>
<TreeItem groupId="A.1" id="A.1.1" isParentExpandedAndVisible={false}>A.1.1</TreeItem>
<TreeItem groupId="A.1.1" isParentExpandedAndVisible={false}>A.1.1.1</TreeItem>
```
Which I don't think is a good DX. What do you think is a better solution to this?

**Comment by diegohaz** [comment]
> I don't think we can. Current useTreeItem depends on `childTreeItems` variable, which is used to determine whether the tree item has child and therefore can be expanded or not. Using shouldRegisterItem means those child will not be registered. So the parent won't be able to determine its expandability.

I see. That makes sense. Using the `disabled` property is fine then.

> Apologize, I updated my previous comment to remove the wrong idea.
> 
> Yes it is indeed a valid state. What I'm trying to say is this: To decide the visibility of A.1.1.1, it needs to know who are its ancestors (A, A.1, A.1.1) and their expanded state beforehand, during SSR. But the data is not available. Using state.items won't be enough, because the items are not registered during SSR. That is what TreeItemSyncContext is trying to solve.
> 
> If we are solving this using props, my assumption is that we are gonna write something like this (prop name "isParentExpandedAndVisible" is just an illustration):
> 
> ```
> const tree = useTreeState({ defaultExpandedIds: ['A.1.1'] });
> ...
> <TreeItem id="A">A</TreeItem>
> <TreeItem groupId="A" id="A.1" isParentExpandedAndVisible={false}>A.1</TreeItem>
> <TreeItem groupId="A.1" id="A.1.1" isParentExpandedAndVisible={false}>A.1.1</TreeItem>
> <TreeItem groupId="A.1.1" isParentExpandedAndVisible={false}>A.1.1.1</TreeItem>
> ```
> 
> Which I don't think is a good DX. What do you think is a better solution to this?

I understand the issue now, and I agree that's not a good API.

This is an issue only with the flat tree because we can't pass data from the parent items to the children through React Context, right? If so, I'd ignore this for now. I expect the flat tree will be used mostly for virtualization, in which case the user will need to pass the items data to the state hook beforehand, so this wouldn't be a problem.

**Comment by natanelia** [comment]
> This is an issue only with the flat tree because we can't pass data from the parent items to the children through React Context, right?

Yes it is

> If so, I'd ignore this for now. I expect the flat tree will be used mostly for virtualization, in which case the user will need to pass the items data to the state hook beforehand, so this wouldn't be a problem.

By ignoring do you mean by let's not support SSR for flat tree for now?

**Comment by diegohaz** [comment]
> By ignoring do you mean by let's not support SSR for flat tree for now?

Yeah

File: packages/ariakit/src/tree/tree-item.tsx [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    props = {
+      role: "treeitem",
+      "aria-expanded": id && hasChildTreeItems ? expanded : undefined,
+      "aria-level": level,
+      "aria-posinset": posInSet,
+      "aria-setsize": setSize,
+      onClick,
+      onKeyDown,
+      ref: useForkRef(ref, props.ref),
+      ...props,
```
Comment (diegohaz) [inline packages/ariakit/src/tree/tree-item.tsx]:
```suggestion
      ...props,
      ref: useForkRef(ref, props.ref),
```

**Comment by diegohaz** [comment]
Thanks for rebasing the PR. Just in case you're wondering, the failing tests on React 17/16 are related to this code:

https://github.com/ariakit/ariakit/blob/1358f5294c2ba24022681378c511059efe118fa7/packages/ariakit/src/collection/collection-state.ts#L97-L102

I'm not sure what to do in this case.

File: packages/ariakit/src/tree/__examples__/tree-flat/test.tsx [hunk 2]
```diff
@@ -5,8 +5,10 @@ import {
   press,
   queryByRole,
   render,
+  waitFor,
 } from "ariakit-test-utils";
 import { axe } from "jest-axe";
+import { act } from "react-dom/test-utils";
```
Comment (diegohaz) [inline packages/ariakit/src/tree/__examples__/tree-flat/test.tsx]: I think you can import `act` from `ariakit-test-utils` as well.

**Comment by natanelia** [comment]
> Thanks for rebasing the PR. Just in case you're wondering, the failing tests on React 17/16 are related to this code:
> 
> https://github.com/ariakit/ariakit/blob/1358f5294c2ba24022681378c511059efe118fa7/packages/ariakit/src/collection/collection-state.ts#L97-L102
> 
> I'm not sure what to do in this case. We can try `requestAnimationFrame` instead of `setTimeout`.

I haven't got much time to debug these issues.
But I have tried using requestAnimationFrame and it makes lots of tests fail.

It does work "most of the time" (flaky) when I use the following:
```
  useEffect(() => {
    const callback = () => setItemsBasedOnDOMPosition(items, setItems);
    const timeout = setTimeout(callback);
    return () => clearTimeout(timeout);
  });
  useEffect(() => {
    setItemsBasedOnDOMPosition(items, setItems);
  }, [items, setItems]);
```

Also I observe some problem using React 16/17 in the website.
1. The first focused item is Item 2 now, indicating the sorting does not trigger properly on first load.
2. The playground disclosure has no initial tab opened.
image

First issue seemed to be solved when I turn off the following check on useSortBasedOnDOMPosition:
```
      // We don't want to sort items if items have been just registered.
      if (!hasPreviousItems) return;
```

Not sure yet what to do. Is there something beside performance optimization wanna be achieved here?

**Comment by diegohaz** [comment]
> I haven't got much time to debug these issues.
> But I have tried using requestAnimationFrame and it makes lots of tests fail.

Hmm! Weird. They shouldn't fail. `setTimeout` works as a debounce there so it's not called multiple times when there are multiple updates at once.

At least on React 18, this doesn't break any of the current tests on main:

```jsx
  useEffect(() => {
    const callback = () => setItemsBasedOnDOMPosition(items, setItems);
    const raf = requestAnimationFrame(callback);
    return () => cancelAnimationFrame(raf);
  });
```


> First issue seemed to be solved when I turn off the following check on useSortBasedOnDOMPosition:
> 
> ```
>       // We don't want to sort items if items have been just registered.
>       if (!hasPreviousItems) return;
> ```
> 
> Not sure yet what to do. Is there something beside performance optimization wanna be achieved here?

Hmm, that check is there because we already check the DOM position when the items get registered:

https://github.com/ariakit/ariakit/blob/de0ea45d5d2d8502d84b3f7c8961fa816fee908b/packages/ariakit/src/collection/collection-state.ts#L145-L146

In theory, we don't need to sort the items again right after they get registered. But maybe we're skipping other runs with that check.

> 2\. The playground disclosure has no initial tab opened.

That's okay. That's an issue on the website, built with React 18 in mind, not on the library. I have an idea of the cause of the problem, but there are more important things to tackle right now.

**Comment by natanelia** [comment]
> Hmm! Weird. They shouldn't fail. setTimeout works as a debounce there so it's not called multiple times when there are multiple updates at once.
At least on React 18, this doesn't break any of the current tests on main

The setTimeout also works fine on React 18 actually. It only fails on React 16 and 17 axe test.

The requestAnimationFrame does work on React 18, and works better than setTimeout on React 16 and 17, but it still fail some Tree tests (not axe ones). And perfectly mimicks this issue:
> The first focused item is Item 2 now, indicating the sorting does not trigger properly on first load.

I'm glad that we now have React 16 and 17 tests though, so we can know this issue beforehand 👍

**Comment by diegohaz** [comment]
@natanelia I guess something went wrong with your rebase.

**Comment by akash07k** [comment]
@natanelia  are you planning to resume the work on this?

**Pull request closed without merging**